tracy-client/src/gpu.rs

```diff
@@ -0,0 +1,171 @@
+use once_cell::sync::Lazy;
+use std::sync::{
+    atomic::{AtomicU16, Ordering},
+    Arc, Mutex,
+};
+
+use crate::Client;
+
+#[repr(u8)]
+///
+pub enum GpuContextType {
+    ///
+    Invalid,
+    ///
+    OpenGL,
+    ///
+    Vulkan,
+    ///
+    OpenCL,
+    ///
+    Direct3D12,
+    ///
+    Direct3D11,
+}
+
+///
+#[derive(Clone)]
+pub struct GpuContext {
+    #[cfg(feature = "enable")]
+    client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    span_index: Arc<AtomicU16>,
+}
+#[cfg(feature = "enable")]
+static GPU_CONTEXT_INDEX: Lazy<Mutex<u8>> = Lazy::new(|| Mutex::new(0));
+
+///
+pub struct GpuSpan {
+    #[cfg(feature = "enable")]
+    _client: Client,
+    #[cfg(feature = "enable")]
+    context: u8,
+    #[cfg(feature = "enable")]
+    // This is the index first query, the second query is always a wrapping_add(1) above this.
+    query_id: u16,
+    #[cfg(feature = "enable")]
+    ended: bool,
+}
+
+impl Client {
+    ///
+    pub fn new_gpu_context(
+        self,
+        name: Option<&str>,
+        ty: GpuContextType,
+        gpu_timestamp: i64,
+        period: f32,
+    ) -> Option<GpuContext> {
+        #[cfg(feature = "enable")]
+        unsafe {
+            // We use a mutex to lock the context index to prevent races when using fetch_add.
+            let mut context_index_guard = GPU_CONTEXT_INDEX.lock().unwrap();
+            if *context_index_guard == 255 {
+                return None;
+            }
+            let context = *context_index_guard;
+            *context_index_guard += 1;
+            drop(context_index_guard);
```

Comment:
This could still use atomics to a great effect, but I’m okay with keeping this as is for the time being too.

Reply:
I don't think we can - you could end up causing UB if your thread timing is really unfortunate.

If the index is 255.

T0 calls atomic_add(1). Before we get to the conditional we're suspended. T1 comes along and tries to increment it, sees that it's value is 0, goes ahead and stomps on the old (still outstanding) gpu context, causing all sorts of nonsense. I think we need to guarantee mutual exclusion when the check is happening.

Reply:
ah `atomic_add` was not what I was thinking of. `compare_exchange` could be used to implement a saturating u8 add, so value could be made to never wrap or exceed `255`.